mopro-core/src/middleware/halo2/mod.rs

```diff
@@ -0,0 +1,122 @@
+use std::collections::HashMap;
+use std::io::Cursor;
+use std::time::Instant;
+
+pub(crate) use halo2_proofs::halo2curves::bn256::{Bn256, Fr as Fp, G1Affine};
+use halo2_proofs::plonk::{ProvingKey, VerifyingKey};
+use halo2_proofs::poly::commitment::Params;
+use halo2_proofs::poly::kzg::commitment::ParamsKZG;
+use halo2_proofs::SerdeFormat::RawBytes;
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+
+use halo2_circuit::{Circuit as TargetCircuit, prove, verify};
+pub use serialisation::deserialize_circuit_inputs;
+
+use crate::MoproError;
+
+mod serialisation;
+
+type CircuitInputs = HashMap<String, Vec<Fp>>;
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct SerializableProof(pub Vec<u8>);
+
+#[derive(Clone, Debug)]
+pub struct SerializablePublicInputs(pub Vec<Fp>);
+
+
+
+/// Read SRS from file
+const SRS_BYTES: &[u8] = include_bytes!(env!("BUILD_SRS_FILE"));
+
+static SRS: Lazy<ParamsKZG<Bn256>> = Lazy::new(|| {
+    let mut reader = Cursor::new(SRS_BYTES);
+    ParamsKZG::read(&mut reader).expect("Unable to read SRS from file")
+});
+
+
+/// Read Proving Key (PK) from file
+
+const PK_BYTES: &[u8] = include_bytes!(env!("BUILD_PK_FILE"));
+
+static PK: Lazy<ProvingKey<G1Affine>> = Lazy::new(|| {
+    let mut reader = Cursor::new(PK_BYTES);
+    ProvingKey::read::<_, TargetCircuit<Fp>>(&mut reader, RawBytes).expect("Unable to read PK from file")
+});
+
+/// Read Verification Key (VK) from file
+
+const VK_BYTES: &[u8] = include_bytes!(env!("BUILD_VK_FILE"));
+
+static VK: Lazy<VerifyingKey<G1Affine>> = Lazy::new(|| {
+    let mut reader = Cursor::new(VK_BYTES);
+    VerifyingKey::read::<_, TargetCircuit<Fp>>(&mut reader, RawBytes).expect("Unable to read VK from file")
+});
+
+
+pub fn generate_halo2_proof2(
```

Comment:
There are two things here:

1. Regarding the suffix _"2"_ - I have used this to indicate similarity to a corresponding Circom method, however, I see that this will rather only cause confusion, hence I will change this.
2. Regarding the suffix _"halo2"_ instead of using `generate_proof` - I am not sure if the interface should be transparent to the proof system used, or rather fail explicitly when compiled for a wrong proof system by using a different function. I would love to hear your views on that, as well as @oskarth

Reply:
In the end, I have renamed the proof and verify functions to be consistent with the circom naming: #173